### DISCUSSION TITLE
MQTT granular topics (breaking scheme for 3.0)

### DISCUSSION THREAD

**Discussion by rcarteraz** [meshtastic/firmware#3798] [Ideas]
Platform
NRF52, ESP32

Description
Currently MQTT is utilized by nodes sending protobuf messages to a channel topic, for example msh/2/c/LongFast. This works and makes sense when using the MQTT broker to link networks via the internet. We could do better for telemetry however. With JSON enabled, nodes could publish sensor data to specific topics, such as msh/2/telemetry/<node id>/battery-voltage. This allows clients to subscribe to the sensor readings they care about without having to parse every message that the node has sent on that channel.

Along with this, an improvement to the telemetry module could be made that enables labelling of sensors. This way if there is multiple temperature sensors hooked up, the user can name them things like internal-temp or cpu-temp and then these can be used in the MQTT topic names.

Originally https://github.com/meshtastic/firmware/issues/2901

**Comment by crypto-smoke** [comment]
I think it would be great if the nodes dumped the packets, as received, direct to MQTT via a topic that contains all the data in the packet header. Namely the node ID of the radio publishing the packet, sender ID, destination ID, and channel hash. This would let users subscribe to packets only destined for their node, or only sent from a specific node. 

An example would be `msh/3/a/1234/5678/4294967295/2` for `msh/<major fw version>/<url format revision (i think)>/<node ID>/<original packet sender ID>/<destination ID>/<channel hash>`

In addition to this, I fully agree that nodes should publish their telemetry as JSON to specific endpoints such as `msh/2/json/<node ID>/voltage`
